### ISSUE TITLE
QueryBuilder - NextGen

### ISSUE THREAD

**Issue by polushinmk** [wp-net/WordPressPCL#22]
I have a suggestion of rewrite the QueryBuilder.
Example of my vision of it:
```
IEnumerable<Post> queryResult=client.Posts.Query()
                       .With(PostQueryEnum.PerPage,20)
                       .With(PostQueryEnum.Author,11)
                       .OrderBy(OrderBy.Date,Order.ASC)
                       .Result();
```
So the main idea is make a query like a pipeline with some common methods:
- With - add a parameter (I think, that use "Where" name is more correct, then "With", but "Where" already used in LINQ and it may cause misunderstanding if we use "Where" too. May be we can use name "Filter" or some another names)
- OrderBy - order results
- Result - execute query and get its result

It will be 1 querybuilder for 1 client type. Posts - PostsQueryBuilder, Comments - CommentsQueryBuilder etc
To limit possible filter values I suggest to create Enums like PostQueryEnum with predefined variables like:
```
enum PostQueryEnum{
     Author,
     PerPage,
     Page,
etc
}
```
I have a working prototype of this QueryBuilder working to query Posts.

Some questions:
- Do you like this?
- We can add more common methods and overloads, ex. overload of With method to apply multiple variables, not only one per call.
```
Dictionary<PostQueryEnum,object> parameters = new Dictionary<PostQueryEnum,object>(){
      new KeyValuePair(PostQueryEnum.PerPage,20),
      new KeyValuePair(PostQueryEnum.Author,1)
}
client.Posts.Query().With(parameters).Result();
```
or like this
```
client.Posts.Query().With((PostQueryEnum.Author,1),(PostQueryEnum.PerPage,20)).Result();
```
Should we do that and which variant more convinient to use?

Looking for your reviews and comments!

**Comment by ThomasPe** [comment]
I agree that there should be one QueryBuilder / Options-Enum for each type so only valid options can be passed.

I would actually prefer it to be closer to your last suggestion, but a bit simpler. I like Linq, but maybe it's a bit too much / generic for this purpose? I had something like this in mind:

```
// either
Dictionary<PostQueryEnum,object> parameters = new Dictionary<PostQueryEnum,object>(){
      new KeyValuePair(PostQueryEnum.PerPage,20),
      new KeyValuePair(PostQueryEnum.Author,1)
}
// or 
PostQuery parameters = new PostQuery(){
    Pages = 20,
    Author = 1
}
// then
var posts = await client.Posts.Query(parameters);
```

I think people using this library will still check the WordPress API Docs to see what parameters can be added to a query, so they should be able to pass all of them in a single object and not wonder if the have to use .Where(), .Filter(), .With()., .OrderBy() etc.

Thoughts?

**Comment by polushinmk** [comment]
Hm...yeah, may be you are right. This task is not too complex to design it with pipeline of Filter(), OrderBy() and Result().
IMO
```
PostQuery parameters = new PostQuery(){
    Pages = 20,
    Author = 1
}
// then
var posts = await client.Posts.Query(parameters);
```
this variant is better then 
```
Dictionary<PostQueryEnum,object> parameters = new Dictionary<PostQueryEnum,object>(){
      new KeyValuePair(PostQueryEnum.PerPage,20),
      new KeyValuePair(PostQueryEnum.Author,1)
}
```
because object creation syntax more easier to use than cumbersome Dictionary and KeyValuePair.
Your current implementation of QueryBuilder looks closer to this variant.
I`ll make some enchancement of it and create another implementations for each client class and then will make a PR to reorg-methods brunch.

**Comment by polushinmk** [comment]
I need feedback on that question:
should we save such methods like 
` public async Task<IEnumerable<Post>> GetPostsByCategory(int categoryId, QueryBuilder builder)` or with new QueryBuilder we dont need them?
methods like `public async Task<IEnumerable<Post>> GetPostsByCategory(int categoryId, bool embed = false)` can be saved, but IMO one Query method can replace all other methods like `public async Task<IEnumerable<Post>> GetPostsByCategory(int categoryId, QueryBuilder builder)` and `public async Task<IEnumerable<Post>> GetPostsByTag(int tagId, QueryBuilder builder)` etc

**Comment by ThomasPe** [comment]
I think we should get rid of all additional methods that take a QueryBuilder-Object. If I create a QueryBuilder-Instance I can just configure everything in it. 

So IMO this might make sense, as it's easier to write:

`public async Task<IEnumerable<Post>> GetPostsByCategory(int categoryId, bool embed = false)`

this does not, since you'll create a QueryBuilder anyway

`public async Task<IEnumerable<Post>> GetPostsByCategory(int categoryId, QueryBuilder builder)`

so you can just use `client.Posts.Query(parameters);`

I'm not against removing all extra methods for V1.0 and re-introducing the most used ones discussing them on a case-by-case basis.

**Comment by ThomasPe** [comment]
merged in 1.0 - lets try it in real applications and collect feedback for V1.1
